scripts/missions/cop/8_1_Garden_of_Antiquity.lua

```diff
@@ -0,0 +1,247 @@
+    mission:setVar(player, ""..towerNum.."-1KILL", 0)
+    mission:setVar(player, ""..towerNum.."-2KILL", 0)
+    mission:setVar(player, ""..towerNum.."-3KILL", 0)
+end
+
+local function clearTowerVars(player)
+    mission:setVar(player, "SouthTower", 0)
+    mission:setVar(player, "WestTower", 0)
+    mission:setVar(player, "EastTower", 0)
+    mission:setVar(player, "SouthTowerCS", 0)
+    mission:setVar(player, "WestTowerCS", 0)
+    mission:setVar(player, "EastTowerCS", 0)
+end
+
+local function setCsVarsFromNpc(player, npc)
+    local npcId = npc:getID()
+    local crystalOffset = npcId - ID.npc.RUBIOUS_CRYSTAL_BASE
+    local cVar = antiquityVars[crystalOffset]
+
+    mission:setVar(player, cVar[2], 1)
+end
+
+local function rubiousCrystalOnTrigger(player, npc)
+    if mission:getVar(player, 'Status') == 1 then
+        local npcId = npc:getID()
+        local crystalOffset = npcId - ID.npc.RUBIOUS_CRYSTAL_BASE
+        local ruaernOffset = ID.mob.RUAERN_BASE + crystalOffset * 3
+        local cVar = antiquityVars[crystalOffset]
+        local thisFightDone = mission:getVar(player, cVar[1])
+        local thisCsAcquired = mission:getVar(player, cVar[2])
+
+        -- spawn ru'aerns
+        if
+            thisCsAcquired == 0 and
+            thisFightDone == 0 and
+            npcUtil.popFromQM(player, npc, { ruaernOffset, ruaernOffset + 1, ruaernOffset + 2 }, { hide = 0 })
+        then
+            clearKillVars(player, crystalOffset)
+            return player:messageSpecial(ID.text.OMINOUS_SHADOW)
+        -- post-fight CS
+        elseif thisCsAcquired == 0 and thisFightDone == 1 then
+            local otherTowerDone1 = mission:getVar(player, cVar[3]) == 1
+            local otherTowerDone2 = mission:getVar(player, cVar[4]) == 1
+
+            if otherTowerDone1 and otherTowerDone2 then
+                return mission:progressEvent(163) -- third CS
+            elseif otherTowerDone1 then
+                return mission:progressEvent(162, cVar[6]) -- second CS. param mentions remaining tower.
+            elseif otherTowerDone2 then
+                return mission:progressEvent(162, cVar[7]) -- second CS. param mentions remaining tower.
+            else
+                return mission:progressEvent(161, cVar[5]) -- first CS. param mentions current tower.
+            end
+        end
+    else
+        return
+    end
+end
+
+mission.sections =
+{
+    {
+        check = function(player, currentMission, missionStatus, vars)
+            return currentMission == mission.missionId
+        end,
+
+        [xi.zone.ALTAIEU] =
+        {
+
+            ['_0x0'] =
+            {
+                onTrigger = function(player, npc)
+                    if mission:getVar(player, 'Status') == 0 then
+                        return mission:progressEvent(164)
+                    end
+                end,
+            },
+
+            ['_0x1'] =
+            {
+                onTrigger = function(player, npc)
```

Comment:
If you define rubiousCrystalOnTrigger as `local rubiousCrystalOnTrigger = function(player, npc)`, then you won't need to wrap it in the function, but instead could do `onTrigger = rubiousCrystalOnTrigger,`

Reply:
Done!